content/blog/2019-10-23-loop-unswitching/index.md

```diff
@@ -0,0 +1,252 @@
+```
+b = True
+for i in range(10):
+    for j in range(10):
+        if b then:
+            do something()
+        else:
+            do something else()
+```
+will not be completely unswitched since the conditional statement lies inside of a nested for loop.  Instead, the conditional will be unswitched outside of the first for loop, which only midly decreases the numbeer of branches.  Implementing this is not too difficult to do, though it requires  additional abstractions for nested for loops.
+
+
+## Results and Evaluation 
+
+##### Results
+The results of this optimization were quite interesting.  I primarily evaluated the code on percentage difference in the number of branches.  To do this, I modified the Bril Typescript interpreter to keep track of the number of branches for each run.  For a program:
+```
+    for _ in range(maxiter):
+        prebody
+        if b:
+            ifbody
+        else:
+            elsebody
+        postbody
+```
+and its unswitched version:
+```
+    if b:
+        for _ in range(maxiter):
+            prebody
+            ifbody
+            postbody
+    else:
+        for _ in range(maxiter):
+            prebody
+            ifbody
+            postbody
+```
+the improvement comes from eliminating branches in the "if...else" decision.  Through testing, one loop unswitching produces a decrease in the number of branches of approximately 14%, on average.  
+
+This result is highly dependent on the structure of the program.  For instance, after unswitching, ```ifbody``` and ```elsebody``` can separately be optimized which would compound the results even further.  The results from the benchmark tests assume that the loops CANNOT be independently optimized, so our claimed average can be thought to be a lower bound on the decrease in the number of branches.
+
+In literature, benchmarking code unswitching involves measuring the increase in code size.  In my implementation, code size roughly doubled for each unswitching which is consistent with the literature.
+
+Overall, this project was a success, and the natural extension is to extend this optimization to work for nested loops.
+
+##### Evaluation 
+
+This optimization was tested on many programs both on defined behavior and undefined behavior.  The testing suite I generated tests on the following (though not limited to) attributes:
+* Should a conditional be unswitched?
+* Simple/Complex blocks before/after conditional
+* Non-unswitched branches before conditional to be unswitched
+* Blocks with no terminating jump instructions
+* Nested if/else conditionals should never be unswitched
+
+I would be very happy with suggestions for additional test cases or pull requests that test this optimization on more test cases.  Furthermore, for questions or comments on design, please reach out by e-mail.
```

Comment:
It's hard to to email you without your email address or some link thereto. :smiley: